From f433ea670f0da0eeed24a035a3f4f355d4a64666 Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Thu, 3 Oct 2024 14:23:07 -0700
Subject: [PATCH] wifi: mac80211: Clear sdata running flag

During wifi down operation, sdata running flag is not cleared which
could cause an issue during recovery. During ieee80211 reconfiguration,
sdata is not expected to be reconfigured during wifi down operation.
As ieee80211 reconfig is scheduled after recovery of a target, this work
won't be aware of any wifi down operation, so it relies on flag to check
if a sdata is required to be reconfigured. The issue is the flag is not
clear then it will have an scenario where one link is going down and
another could be going up for the same sdata. It is expected to honor
wifi down operation when a down operation is requested.

Fix by clearing the running flag during an interface down operation.

In addition during any down operation,
currently, in this subroutine compat variable is assiged to NULL by
default. The code checks for compatability on cfg80211 chandef for all
interfaces exist at the time of assignment of link chanctx.

The issue arises when not all interfaces are running or are
of type AP_VLAN, triggering the WARN_ON_ONCE(!compat).

Fix by removing the WARN_ON_ONCE for !compact check under these
conditions, as it is more likely to be encountered.

Patch-dependency: None
Patch-work: None

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 net/mac80211/cfg.c | 1 +
 1 file changed, 1 insertion(+)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1840,6 +1840,7 @@ static int ieee80211_stop_ap(struct wiph
 	link_conf = link->conf;
 	sdata_assert_lock(sdata);
 
+	clear_bit(SDATA_STATE_RUNNING, &sdata->state);
 	old_beacon = sdata_dereference(link->u.ap.beacon, sdata);
 	if (!old_beacon)
 		return -ENOENT;
--- a/net/mac80211/chan.c
+++ b/net/mac80211/chan.c
@@ -843,7 +843,7 @@ void ieee80211_recalc_chanctx_chantype(s
 		}
 	}
 
-	if (WARN_ON_ONCE(!compat)) {
+	if (!compat) {
 		rcu_read_unlock();
 		return;
 	}
